giraffe/src/utils/fluxParsing.ts

```diff
@@ -0,0 +1,149 @@
+import Papa from 'papaparse'
+import uuid from 'uuid'
+
+import {FluxTable} from '../types'
+
+import {get} from './get'
+import {groupBy} from './groupBy'
+
+export const parseResponseError = (response: string): FluxTable[] => {
+  const data = Papa.parse(response.trim()).data as string[][]
+
+  return [
+    {
+      id: uuid.v4(),
+      name: 'Error',
+      result: '',
+      groupKey: {},
+      dataTypes: {},
+      data,
+    },
+  ]
+}
+
+/*
+  A Flux CSV response can contain multiple CSV files each joined by a newline.
+  This function splits up a CSV response into these individual CSV files.
+
+  See https://github.com/influxdata/flux/blob/master/docs/SPEC.md#multiple-tables.
+*/
+export const parseChunks = (response: string): string[] => {
+  const trimmedResponse = response.trim()
+
+  if (trimmedResponse === '') {
+    return []
+  }
+
+  // Split the response into separate chunks whenever we encounter:
+  //
+  // 1. A newline
+  // 2. Followed by any amount of whitespace
+  // 3. Followed by a newline
+  // 4. Followed by a `#` character
+  //
+  // The last condition is [necessary][0] for handling CSV responses with
+  // values containing newlines.
+  //
+  // [0]: https://github.com/influxdata/influxdb/issues/15017
+
+  const chunks = trimmedResponse
+    .split(/\n\s*\n#/)
+    .map((s, i) => (i === 0 ? s : `#${s}`))
+
+  return chunks
+}
+
+export const parseResponse = (response: string): FluxTable[] => {
+  const chunks = parseChunks(response)
+  const tables = chunks.reduce((acc, chunk) => {
+    return [...acc, ...parseTables(chunk)]
+  }, [])
+
+  return tables
+}
+
+export const parseTables = (responseChunk: string): FluxTable[] => {
+  const lines = responseChunk.split('\n')
+  const annotationLines: string = lines
+    .filter(line => line.startsWith('#'))
+    .join('\n')
+    .trim()
+  const nonAnnotationLines: string = lines
+    .filter(line => !line.startsWith('#'))
+    .join('\n')
+    .trim()
+
+  if (!annotationLines) {
+    throw new Error('Unable to extract annotation data')
+  }
+
```

Comment:
```suggestion
  const nonAnnotationLines: string = lines
    .filter(line => !line.startsWith('#'))
    .join('\n')
    .trim()
```

Reply:
suggestion: see if we can early return before creating a new variable

Reply:
Done.